pkg/server/server.go

```
@@ -222,15 +246,161 @@ func (s *Server) Run(ctx context.Context) error {
+			// server handle it.
+			crdName := requestInfo.Resource + ".core"
+			_, err := serverChain.CustomResourceDefinitions.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().GetWithContext(ctx, crdName)
+			if err == nil {
+				serverChain.CustomResourceDefinitions.GenericAPIServer.Handler.NonGoRestfulMux.ServeHTTP(res.ResponseWriter, req.Request)
+				return
+			}
+		} else if req.Request.URL.Path == "/api/v1" || req.Request.URL.Path == "/api/v1/" {
+			// This is a discovery request. We may need to combine discovery from the GenericControlPlane (which has built-in v1 types)
+			// and CRDs, if there are any v1 CRDs.
+
+			// If we are retrieving the GenericControlPlane's v1 APIResources, let that request through.
+			const passthroughHeader = "X-Kcp-Api-V1-Discovery-Passthrough"
+			if _, passthrough := req.Request.Header[passthroughHeader]; passthrough {
+				chain.ProcessFilter(req, res)
+				return
+			}
+
+			// Get all the CRDs (in the context's logical cluster) to see if any of them are in v1
+			crds, err := serverChain.CustomResourceDefinitions.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().ListWithContext(ctx, labels.Everything())
+			if err != nil {
+				// FIXME
+			}
+			crdDiscovery, ok := apiextensionsapiserver.APIResourcesForGroupVersion("", "v1", crds)
+			if !ok {
+				// No v1 CRDs - let the request through.
+				chain.ProcessFilter(req, res)
+				return
+			}
+
+			// v1 CRDs present - need to clone the request, add our passthrough header, and get /api/v1 discovery from
+			// the GenericControlPlane's server.
+			cr := utilnet.CloneRequest(req.Request)
+			cr.Header.Add(passthroughHeader, "1")
+
+			writer := newInMemoryResponseWriter()
+			serverChain.GenericControlPlane.GenericAPIServer.Handler.GoRestfulContainer.ServeHTTP(writer, cr)
+			if writer.respCode != http.StatusOK {
+				// FIXME
```

No error handling here (yet)... this means the attempt to get /api/v1 for built-in types failed. Probably just need to echo the response back?

---

that's a "should never happen" case, right? Would fail hard here, i.e. pass up that error to the real request.

---

Agreed